static/js/chart/draw.ts

```diff
@@ -427,10 +434,207 @@ function drawLineChart(
+      .scaleTime()
+      .domain(d3.extent(dataGroups[0].value, (d) => new Date(d.label).getTime()))
+      .range([MARGIN.left, width - MARGIN.right - legendWidth]);
+
+  let yScale = d3
+      .scaleLinear()
+      .domain([minV, maxV])
+      .range([height - MARGIN.bottom, MARGIN.top])
+      .nice(NUM_Y_TICKS);
+
+  addXAxis(svg, height, xScale);
+  addYAxis(svg, width, yScale, unit);
+
+  let dashIndex = 0;
+  for (let geoId in dataGroupsDict) {
+    dataGroups = dataGroupsDict[geoId];
+    for (let i = 0; i < dataGroups.length; i++) {
+      let dataGroup = dataGroups[i];
+      let dataset = dataGroup.value.map(function (dp) {
+        return [new Date(dp.label).getTime(), dp.value];
+      });
+
+      let line = d3
+          .line()
+          .x((d) => xScale(d[0]))
+          .y((d) => yScale(d[1]));
+
+      svg
+          .append("path")
+          .datum(dataset)
+          .attr("class", "line")
+          .style("stroke", styles["colors"][i])
+          .attr("d", line)
+          .attr("stroke-width", "2")
+          .attr("stroke-dasharray", styles["dashes"][dashIndex]);
+    }
+
+    dashIndex++;
+  }
+
+  let legendId = randDomId();
+  svg
+      .append("g")
+      .attr("id", legendId)
+      .attr("transform", `translate(${width - legendWidth}, 50)`);
+
+  buildLegend(legendId, styles);
+
+  // return colors here used to add menu below the chart.
+  return styles["colors"];
+}
+
+
+/**
+ * Generate in-chart legend.
+ *
+ * @param id: This is the id for the chart legend element.
+ * @param styles: It contains all colors and dashes for geoIds and statVars.
+ */
+function buildLegend(
+    id: string,
+    styles: {},
+) {
+  let legend = d3
+      .select("#" + id);
+
+  if (styles["dashes"].length == 1) {
+    // Only have one geoId. Then different statsVars should have different colors.
+    let index = 0;
+    for (let color of styles["colors"]) {
+      let legendClass = legend
+          .append("g")
+          .attr("transform", `translate(0, ${MARGIN.legendHeight * index})`);
+
+      legendClass
+          .append("text")
+          .attr("x", "40")
+          .attr("y", `${MARGIN.legendHeight * index}`)
+          .text(styles["statVars"][index])
+          .style("font-size", "14")
+          .attr("fill", `${color}`);
+
+      index++;
+    }
+  } else {
+    // Have multiply goeIds. Then different geoIds should have different dashes.
+    let index = 0;
+    for (let geo of styles["geoIds"]) {
+      let legendClass = legend
+          .append("g")
+          .attr("transform", `translate(0, ${MARGIN.legendHeight * index})`);
+
+      legendClass
+          .append("line")
+          .attr("stroke-width", 2)
+          .attr("x1", "0")
+          .attr("y1", `${MARGIN.legendHeight * index - MARGIN.legendAdjustment}`)
+          .attr("x2", "30")
+          .attr("y2", `${MARGIN.legendHeight * index - MARGIN.legendAdjustment}`)
+          // Default color to be black here.
+          .attr("stroke", "#000000")
```

Comment:
Is the default stroke color required? It might already be black if you don't specify this.

Reply:
Seems originally it is white, as I can't see anything in a white background. So I specify the black color here.

Reply:
Great, thanks for the explanation!